MM-10924 Add forgot password screen

sudheerDev opened the pull request:
#### Summary
 * Add a new screen for the user to reset password.
   but can only verify the link through web UI because of few constraints as universal linking.

#### Ticket Link
[MM-10924](https://mattermost.atlassian.net/browse/MM-10924)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: [ios emulator] 

#### Screenshots
### login screen
<img width="366" alt="screen shot 2018-07-04 at 4 25 47 pm" src="https://user-images.githubusercontent.com/4973621/42273306-728c37a0-7fa6-11e8-8af7-f9a337baaafb.png">

### Reset password screen
<img width="367" alt="screen shot 2018-07-04 at 4 25 53 pm" src="https://user-images.githubusercontent.com/4973621/42273316-7ab3ab0c-7fa6-11e8-8e97-e9e400a326e6.png">

### success message
<img width="354" alt="screen shot 2018-07-04 at 4 26 04 pm" src="https://user-images.githubusercontent.com/4973621/42273392-a86c51de-7fa6-11e8-80a5-ce480db3fff3.png">

sudheerDev commented on assets/base/i18n/en.json:
```diff
@@ -2707,7 +2707,7 @@
   "password_send.email": "Email",
   "password_send.error": "Please enter a valid email address.",
-  "password_send.link": "If the account exists, a password reset email will be sent to: <br/><b>{email}</b><br/><br/>",
+  "password_send.link": "If the account exists, a password reset email will be sent to:",
```
the html tags cause issues with rendering formatted text. So I had to remove it. Let me know someone has a better solution for this.

saturninoabril replied:
I think you need to leave it as is and just create another one specifically for mobile RN (starting with `mobile.`).  Also, you might want to be the first to implement `FormattedMarkdownText`.

sudheerDev replied:
tried `FormattedMarkdownText` this but it keeps throwing weird errors. so added a new translation key for mobile.

sudheerDev replied:
error is `invariant-violation-view-config-not-found-for-name-span` which did not help all that much. I will dig into to more and see if i can fix it.

sudheerDev replied:
@hmhealey I added a new key for mobile and reverted web key as before so I guess this should be good.

sudheerDev replied:
Hey @enahum I raised a [PR](https://github.com/mattermost/mattermost-webapp/pull/1419) to change the key in webapp. Once it is merged should I update it here as well?

sudheerDev commented on app/utils/general.js:
```diff
@@ -69,3 +69,13 @@ export function throttle(fn, limit, ...args) {
     };
 }
+
+export function isEmail(email) {
```
taken from webapp utils.

hmhealey replied:
We might want to move this into `mattermost-redux` so that both apps can share code for it

sudheerDev commented on app/screens/forgot_password/forgot_password.js:
```diff
@@ -0,0 +1,200 @@
+    resetSuccessContainer: {
+        marginTop: 15,
+        padding: 10,
+        backgroundColor: '#dff0d8',
```
colours and backgrounds are taken from webapp. they are of boot-strap classes and so no theming exists

saturninoabril replied:
Please verify that it's working as expected when changing user theme.

sudheerDev replied:
I think we don't support themes on login and password reset pages so this should be fine.

saturninoabril commented on app/screens/forgot_password/forgot_password.js:
```diff
@@ -0,0 +1,200 @@
+    };
+
+    blur = () => {
+        if (this.email) {
```
Should this be `this.emailId`?

saturninoabril commented on app/screens/forgot_password/forgot_password.test.js:
```diff
@@ -0,0 +1,78 @@
+        expect(actions.sendPasswordResetEmail).toBeCalledWith('test@test.com');
+    });
+
+    test('match smapshot after success of sendPasswordResetEmail', async () => {
```
smapshot -> snapshot

saturninoabril commented on app/screens/login/login.js:
```diff
@@ -394,6 +411,16 @@ export default class Login extends PureComponent {
+                            <FormattedText
+                                id='login.forgot'
+                                defaultMessage='I forgot my password'
+                                style={{color: '#2389D7'}}
```
Maybe add `{color: '#2389D7'}` to style constant?

sudheerDev commented:
@saturninoabril addressed review comments

hmhealey commented on app/screens/forgot_password/forgot_password.js:
```diff
@@ -0,0 +1,200 @@
+        if (this.emailId) {
+            this.emailId.blur();
+        }
+        Keyboard.dismiss();
```
I don't think this is necessary since the keyboard should be dismissed automatically when the input loses focus.

hmhealey commented on assets/base/i18n/en.json:
```diff
@@ -2495,6 +2495,7 @@
   "mobile.offlineIndicator.offline": "Cannot connect to the server",
   "mobile.open_dm.error": "We couldn't open a direct message with {displayName}. Please check your connection and try again.",
   "mobile.open_gm.error": "We couldn't open a group message with those users. Please check your connection and try again.",
+  "mobile.password_send.link": "If the account exists, a password reset email will be sent to:",
```
As mentioned in my other comment, we could change the plain `password_send.link` to remove the HTML from it since it's something that can be translated and understood on its own so that we can also use it here, but that would need to be done in the web app repo first.

sudheerDev commented:
@hmhealey Fixed the review comments 
 *Moved isEmail to redux lib. 
 * Updated redux version. 
 * Removed the newly added en.json key and raised a PR to modify in webapp(I modified the existing key here as well for now. I will revert it if needed once I get to know that change here is not needed anymore.).

sudheerDev commented:
@enahum Hey, I have a [string](https://github.com/mattermost/mattermost-mobile/blob/123e0e699f90a6d98019f4c03e3ff2032f7e586b/assets/base/i18n/en.json#L2730) change in the webapp PR. I don't see the change in existing en.json. Should I be making the same change here as well?

enahum commented:
@sudheerDev do not worry about it we haven’t made the PR for the mobile app that contains the changes made in the en.json file from the webapp just yet, but you should be good if you added the right id and defaultMessage, then once the translation PR is merged then the en.json file will have the string
